backend/src/apiserver/server/visualization_server.go

```diff
@@ -115,13 +112,12 @@ func (s *VisualizationServer) getVisualizationServiceURL(request *go_client.Crea
 	return s.serviceURL
 }
 
-func isVisualizationServiceAlive(serviceURL string) bool {
-	resp, err := http.Get(serviceURL)
-	if err != nil {
-		glog.Error("Unable to verify visualization service is alive!", err)
-		return false
+func isVisualizationServiceAlive(serviceURL string) (bool, error) {
+	resp, _ := http.Get(serviceURL)
```

Comment:
So I did some investigation and what happened here in `TestGenerateVisualization_ServiceNotAvailableError`:
```
resp, err := http.Get(serviceURL) # resp.StatusCode != http.StatusOK and err == nil
```
So I don't think we can simply use:

```
return false, util.Wrap(err, "ERROR MESSAGE HERE") 
```

Because `util.Wrap(err, "ERROR MESSAGE HERE") ` will be `nil` if `err == nil`

Reply:
Good point! I didn't know this.

Based on this new information, then error can come from:
* `http.Get`'s err -> network errors etc (you can read http.Get documentation for when error shows up), detailed root cause can be found at `err`
* request succeeded but resp.statusCode is not ok -> resp.status and resp.body might contain information about why it failed

So I think it'll be better if we can propagate error for both cases:
* when there is an error, we can use `util.Wrap(err, '....')`
* when statuscode not OK, we can return `resp.status` inside the error because resp.status should be short
  and log `resp.body` using `glog.Error`, because it can be long.

The benefit will be when there's an error, we can get all the failure information without adding rebuilding api server and adding logging there.

Reply:
Thanks for providing more insights! I made more investigations and some changes to address your comments.



The lifecycle of `TestGenerateVisualization_ServiceNotAvailableError` unit test:

1. Create necessary resources including `server `and `request`.
2. Then make a call `generateVisualizationFromRequest`: when we reach out to `resp, err := http.Get(serviceURL)`, in our case, it will return non-nil `resp` and `err == nil` when get a `non-2xx response (It's 500 in test)`
3. So the `resp.Status` will always be `500 Internal Server Error `.



Thus, users will be able to see similar logs as below when vis-server is not accessible:

```
Unable to verify visualization service is alive by sending request to http://127.0.0.1:53545 and get 500 Internal Server Error !
```

Reply:
Thanks, the case coverred in unit test is good now.

Would you mind adding handling for
> http.Get's err -> network errors etc (you can read http.Get documentation for when error shows up), detailed root cause can be found at err

?

I think you can reproduce this case by connecting to a non existant host in unit test.

Reply:
I'm not sure if I understand correctly, from my understanding, what you want:

In unit test, we create `server` with non-existing host, and then `http.Get()` will return `err != nil`. Can you correct me if I'm wrong?

But I'm not sure if this is the expected behavior for `TestGenerateVisualization_ServiceNotAvailableError ` test, or you're talking about adding another unit test to handle the behavior you mentioned:

> http.Get's err -> network errors etc (you can read http.Get documentation for when error shows up), detailed root cause can be found at err

Reply:
> In unit test, we create server with non-existing host, and then http.Get() will return err != nil

Correct!

> or you're talking about adding another unit test to handle the behavior you mentioned

Exactly, this should be a new unit test.

Reply:
Thanks for explanations! I added another unit test to handle `ServerHostNotExistError`.